### ISSUE TITLE
About path length

### ISSUE THREAD

**Issue by only-changer** [wouterkool/attention-learn-to-route#9]
Hi,
I have read your paper and find that in the TSP problem all paths have the same length, so I wonder can your method works when the length of paths isn't fixed?

**Comment by wouterkool** [comment]
Hi!

The model architecture is not tied to a specific path length / graph size, so you can run with any sized input.

If you want to have different sizes of input graphs in a *single* batch, this is *not* currently possible. However, you can run them one by one with a batch size of one.

Alternatively, you could pad all graphs with 'dummy nodes' to make them the same size, and implement an additional mask in the different attention operations, and you will have to deal with outputs of different lengths (similar to some of the VRP problems, for example).

You can verify if your implementation is correct by comparing against running one at a time: the results should be the same (at least when batch norm and randomness is disabled, e.g. at greedy test time).

**Comment by only-changer** [comment]
I don't want to change the size of graphs, I am just wondering how to output path with different lengths on the same size graph. For example, for a fixed graph, the shortest path of two nodes may have different lengths. Can you decoder work in this situation? 
I know that we can add dummy nodes on the path to make them all have the same length, but is there a better way?

**Comment by wouterkool** [comment]
I'm not sure if I fully understand it. This repository solves TSP like problems, computing an order of nodes. If you want to find the shortest paths between two nodes, that's a different problem for which I would recommend Dijkstra or A*. You can still try to modify this repository to solve that problem, however.

As regards to the different lengths, I think your options are using a batch size of 1, using a larger batch size with padding or using something which dynamically reduces the batch size if some sequences are finished. This last option is most memory efficient and somewhat similar to [PackedSequence](https://pytorch.org/docs/stable/nn.html?highlight=packedsequence#torch.nn.utils.rnn.PackedSequence)s, but a bit more difficult to implement as the sequences are not given upfront.

**Comment by only-changer** [comment]
ok, I got it. Thanks!
